src/widgets/CodeView.js

```diff
@@ -81,6 +147,74 @@ class CodeView extends Gtk.Widget {
     hover.add_provider(provider);
   }
 
+  #onCompletionRequest = (provider, request) => {
+    console.log(`completion-request: ${request.context.get_word()}`);
+
+    const [, start, end] = request.context.get_bounds();
+    const text = this.buffer.get_text(start, end, false);
+
+    // log(request.context.get_proposals_for_provider(completion_provider));
+
+    // console.log({
+    //   start: {
+    //     line: start.get_line(),
+    //     offset: start.get_line_offset(),
+    //   },
+    //   end: {
+    //     line: end.get_line(),
+    //     offset: end.get_line_offset(),
+    //   },
+    //   text,
+    // });
+
+    this.css
+      .completion(end)
+      .then((result) => {
+        log(result);
+
+        result.forEach((completion) => {
+          if (completion.insertText?.startsWith(text)) {
+            request.add(new Proposal(completion));
+          }
+        });
+      })
+      .catch(logError)
+      .finally(() => {
+        request.state_changed(Workbench.RequestState.COMPLETE);
```

Comment:
I'm honestly not sure what's going wrong, since I'm not familiar with GtkSource. I noticed the `finally()` handler is not getting called very often though.